core/src/main/java/org/pentaho/platform/workitem/WorkItemLifecycleEvent.java

```diff
@@ -0,0 +1,222 @@
+
+  private String workItemUid;
+  private String workItemDetails;
+  private WorkItemLifecyclePhase workItemLifecyclePhase;
+  private String lifecycleDetails;
+  private Date sourceTimestamp;
+  private Date targetTimestamp;
+  private String sourceHostName;
+  private String sourceHostIp;
+  private static String HOST_NAME;
+  private static String HOST_IP;
+
+  static {
+    try {
+      HOST_NAME = InetAddress.getLocalHost().getCanonicalHostName();
+      HOST_IP = InetAddress.getLocalHost().getHostAddress();
+    } catch ( final UnknownHostException uhe ) {
+      logger.error( uhe.getLocalizedMessage() );
+    }
+  }
+
+  /**
+   * Default constructor, needed for serialization purposes.
+   */
+  protected WorkItemLifecycleEvent() {
+  }
+
+  /**
+   * Creates the {@link WorkItemLifecycleEvent} with all the required parameters.
+   *
+   * @param workItemUid            a {@link String} containing unique identifier for the {@link WorkItemLifecycleEvent}
+   * @param workItemDetails        a {@link String} containing details of the {@link WorkItemLifecycleEvent}
+   * @param workItemLifecyclePhase a {@link WorkItemLifecyclePhase} representing the lifecycle event
+   * @param lifecycleDetails       a {@link String} containing any additional details about the lifecycle event, such as
+   *                               pertinent failure messages
+   * @param sourceTimestamp        a {@link Date} representing the time the lifecycle change occurred.
+   */
+
+  public WorkItemLifecycleEvent( final String workItemUid, final String workItemDetails, final WorkItemLifecyclePhase
+    workItemLifecyclePhase, final String lifecycleDetails, final Date sourceTimestamp ) {
+    this.workItemUid = workItemUid;
+    this.workItemDetails = workItemDetails;
+    this.workItemLifecyclePhase = workItemLifecyclePhase;
+    this.lifecycleDetails = lifecycleDetails;
+    this.sourceTimestamp = sourceTimestamp;
+
+    // if the workItemUid is null, generate it
+    if ( StringUtil.isEmpty( this.workItemUid ) ) {
+      this.workItemUid = UUID.randomUUID().toString();
+    }
+    // Set sourceTimestamp to current date only if not already provided
+    if ( this.sourceTimestamp == null ) {
+      this.sourceTimestamp = new Date();
+    }
+    // target timestamp is always a fresh date, set at the time the event object is created; this is because the
+    // original event may be created on a different host, where  the clock is off; the target timestamp is guaranteed
+    // to be expressed in terms of a common clock on the target host, where the event is actually persisted
+    this.targetTimestamp = new Date();
+
+    // set the default values for host name and ip, they can be changed directly if needed
+    this.sourceHostName = HOST_NAME;
+    this.sourceHostIp = HOST_IP;
+  }
+
+  /**
+   * Looks up the {@code ActionUtil.REQUEST_ID} within the {@link Map}. If available, the value is returned,
```

Comment:
Can we generate a report of all the events a work item went through? I'm thinking (long run) an admin running a report of all the work items processed during a day with various criteria like work items that failed or couldn't run or generated a particular error message to see a pattern and adjust their cluster resources etc. It would be good if we could tie all the WorkItemEvents to a WorkItem.